ui/decredmaterial/tabs.go

```diff
@@ -1,71 +1,306 @@
+
+	layout.Stack{}.Layout(gtx,
+		layout.Stacked(func() {
+			t.iconText(gtx, tabPosition)()
+			if gtx.Dimensions.Size.X > adaptiveTabWidth {
+				adaptiveTabWidth = gtx.Dimensions.Size.X
+			}
+		}),
+		layout.Expanded(func() {
+			if tabPosition == Left || tabPosition == Right {
+				gtx.Constraints.Width.Min = adaptiveTabWidth
+			}
+			Button{Background: color.RGBA{}, Radius: 0, shaper: font.Default()}.Layout(gtx, btn)
+			tabWidth, tabHeight = tabIndicatorDimensions(gtx, tabPosition)
+		}),
+		layout.Expanded(func() {
+			if selected != t.index {
+				return
+			}
+			indicatorDirection(tabPosition).Layout(gtx, func() {
+				layout.Flex{Axis: layout.Horizontal}.Layout(gtx,
+					layout.Rigid(line(gtx, tabWidth, tabHeight, keyblue)))
+			})
+		}),
+	)
+}
+
+// Tabs displays succession of TabItems. Using the Position option, Tabs can be displayed on any four sides
+// of a rendered page.
 type Tabs struct {
-	List     *layout.List
-	Flex     layout.Flex
-	Size     float32
-	items    []TabItem
-	Selected int
-	changed  bool
-	btns     []*widget.Button
+	flex        layout.Flex
+	Size        float32
+	items       []TabItem
+	Selected    int
+	changed     bool
+	btns        []*widget.Button
+	list        *layout.List
+	Position    Position
+	Separator   bool
+	scrollLeft  *widget.Button
+	scrollRight *widget.Button
 }
 
 func NewTabs() *Tabs {
 	return &Tabs{
-		List: &layout.List{
-			Axis: layout.Vertical,
-		},
-		Size: DefaultTabSize,
+		list:        &layout.List{},
+		Position:    Left,
+		scrollLeft:  new(widget.Button),
+		scrollRight: new(widget.Button),
 	}
 }
 
-func (tab *Tabs) SetTabs(tabs []TabItem) {
-	tab.items = tabs
-	if len(tab.items) != len(tab.btns) {
-		tab.btns = make([]*widget.Button, len(tab.items))
-		for i := range tab.btns {
-			tab.btns[i] = new(widget.Button)
+// SetTabs creates a button widget for each tab item.
+func (t *Tabs) SetTabs(tabs []TabItem) {
+	t.items = tabs
+	if len(t.items) != len(t.btns) {
+		t.btns = make([]*widget.Button, len(t.items))
+		for i := range t.btns {
+			t.btns[i] = new(widget.Button)
 		}
 	}
 }
 
-func (tab *Tabs) Changed() bool {
-	return tab.changed
+// Changed returns the changed state of the tab.
+func (t *Tabs) Changed() bool {
+	return t.changed
 }
 
-// Layout the tabs
-func (tab *Tabs) Layout(gtx *layout.Context, body layout.Widget) {
-	tab.Flex.Layout(gtx,
-		layout.Flexed(tab.Size, func() {
-			tab.List.Layout(gtx, len(tab.btns), func(i int) {
-				tab.items[i].Layout(gtx, tab.btns[i], i == tab.Selected)
-			})
-		}),
-		layout.Flexed(1-tab.Size, body),
-	)
-	for i := range tab.btns {
-		tab.changed = false
-		if tab.btns[i].Clicked(gtx) {
-			tab.changed = true
-			tab.Selected = i
+// scrollButton lays out the right and left scroll buttons of the tab when Position is Horizontal.
+func (t *Tabs) scrollButton(gtx *layout.Context, right bool, button *widget.Button) layout.FlexChild {
+	show := false
+	icon := mustIcon(NewIcon(icons.NavigationChevronLeft))
```

Comment:
Decoding the icon should be done before the icon needs to be laid out and done only once.